File: chain/python/mychain.py
from abc import ABC, abstractmethod

class Chef(ABC):
    def __init__(self):
        self.nextchef = None

    @abstractmethod
    def setNextChef(self, chef):
        pass

    @abstractmethod
    def cook(self):
        pass


class BasicChef(Chef):
    def __init__(self):
        super().__init__()

    def setNextChef(self, chef):
        self.nextchef = chef
        return self.nextchef

    def cook(self):
        print("Basic Chef Cooking...")
        if self.nextchef:
            self.nextchef.cook()


class CollectingIngredientsChef(Chef):
    def __init__(self):
        super().__init__()

    def setNextChef(self, chef):
        self.nextchef = chef
        return self.nextchef

    def cook(self):
        print("Collecting Ingredients Chef Cooking...")
        if self.nextchef:
            self.nextchef.cook()


class BoilingChef(Chef):
    def __init__(self):
        super().__init__()

    def setNextChef(self, chef):
        self.nextchef = chef
        return self.nextchef

    def cook(self):
        print("Boiling Chef Cooking...")
        if self.nextchef:
            self.nextchef.cook()
            
class FryingChef(Chef):
    def __init__(self):
        super().__init__()

    def setNextChef(self, chef):
        self.nextchef = chef
        return self.nextchef

    def cook(self):
        print("Frying Chef Cooking...")
        if self.nextchef:
            self.nextchef.cook()
        
class MasterChef(Chef):
    def __init__(self):
        super().__init__()

    def setNextChef(self, chef):
        self.nextchef = chef
        return self.nextchef

    def cook(self):
        print("Master Chef Cooking...")
        if self.nextchef:
            self.nextchef.cook()

def main():
    basicChef = BasicChef()
    basicChef.cook()
    
    collectingIngredientsChef = CollectingIngredientsChef()
    boilingChef = BoilingChef()
    fryingChef = FryingChef()
    masterChef = MasterChef()

    basicChef.setNextChef(collectingIngredientsChef) \
        .setNextChef(boilingChef).setNextChef(fryingChef) \
        .setNextChef(masterChef)
    basicChef.cook()


if __name__ == "__main__":
    main()

File: chain/swift/mychain.swift
import Foundation

// The Chef protocol defining the Chain of Responsibility interface
protocol Chef: AnyObject {
    var nextChef: Chef? { get set }
    func setNext(chef: Chef) -> Chef
    func cook(dish: String)
}

// Extension to provide default implementation for linking
extension Chef {
    func setNext(chef: Chef) -> Chef {
        self.nextChef = chef
        return chef
    }
}

// 1. Collecting Ingredients Chef
class CollectingIngredientsChef: Chef {
    var nextChef: Chef?
    
    func cook(dish: String) {
        print("CollectingIngredientsChef: Gathering fresh ingredients for \(dish).")
        nextChef?.cook(dish: dish)
    }
}

// 2. Boiling Chef
class BoilingChef: Chef {
    var nextChef: Chef?
    
    func cook(dish: String) {
        if dish.lowercased().contains("soup") || dish.lowercased().contains("pasta") {
            print("BoilingChef: Boiling water and preparing the base for \(dish).")
        }
        nextChef?.cook(dish: dish)
    }
}

// 3. Frying Chef
class FryingChef: Chef {
    var nextChef: Chef?
    
    func cook(dish: String) {
        if dish.lowercased().contains("fries") || dish.lowercased().contains("steak") {
            print("FryingChef: Searing and frying components for \(dish).")
        }
        nextChef?.cook(dish: dish)
    }
}

// 4. Basic Chef
class BasicChef: Chef {
    var nextChef: Chef?
    
    func cook(dish: String) {
        print("BasicChef: Plating and garnishing \(dish).")
        nextChef?.cook(dish: dish)
    }
}

// 5. Master Chef (The Final Authority)
class MasterChef: Chef {
    var nextChef: Chef?
    
    func cook(dish: String) {
        print("MasterChef: Final quality check. \(dish) is ready for the guest!\n")
    }
}

// Implementation in "Main"
let ingredientsChef = CollectingIngredientsChef()
let boilingChef = BoilingChef()
let fryingChef = FryingChef()
let basicChef = BasicChef()
let masterChef = MasterChef()

// Setting up the chain: Ingredients -> Boiling -> Frying -> Basic -> Master
ingredientsChef
    .setNext(chef: boilingChef)
    .setNext(chef: fryingChef)
    .setNext(chef: basicChef)
    .setNext(chef: masterChef)

// Execute the chain
print("--- Ordering Tomato Soup ---")
ingredientsChef.cook(dish: "Tomato Soup")

print("--- Ordering Fried Steak ---")
ingredientsChef.cook(dish: "Fried Steak")
